src/mlpack/methods/ann/layer/cross_entropy_error_impl.hpp

```diff
@@ -0,0 +1,59 @@
+/**
+ * @file cross_entropy_error_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of the cross-entropy performance function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_CROSS_ENTROPY_ERROR_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_CROSS_ENTROPY_ERROR_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "cross_entropy_error.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+CrossEntropyError<InputDataType, OutputDataType>::CrossEntropyError(double eps)
+  : eps(eps)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+double CrossEntropyError<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, const arma::Mat<eT>&& target)
+{
+  return -arma::accu(target % arma::log(input + eps) +
+                     (1. - target) % arma::log(1. - input + eps));
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void CrossEntropyError<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& input,
+    const arma::Mat<eT>&& target,
+    arma::Mat<eT>&& output)
+{
+  output = (1. - target) / (1. - input + eps) - target / (input + eps);
```

Comment:
Checked it on a piece of paper and got the precise expression we should get after simplifying the original expression so that it would still be the gradient of the loss function with `log(x + eps)`:

```output = (input - target + eps * (1. - 2 * target)) / ((1. - input + eps) % (input + eps))```

This one, however, doesn't really optimize much (or, at least, I think so), because even though it runs only one (element-wise) division, it runs *three* multiplications, which are also slow (as compared with additions - that's why I didn't count them).